pennylane/ops/functions/bind_new_parameters.py

```diff
@@ -0,0 +1,165 @@
+    else:
+        invalid_params = len(params) != op.num_params
+
+    if invalid_params:
+        raise ValueError(
+            "The length of the new parameters does not match the expected shape; "
+            f"got {len(params)}, expected {op.num_params}."
+        )
+
+    return params
+
+
+@singledispatch
+def bind_new_parameters(op: Operator, params: Optional[Sequence[TensorLike]] = None) -> Operator:
+    """Create a new operator with updated parameters
+
+    This function takes an :class:`~.Operator` and new parameters as input and
+    returns a new operator of the same type with the new parameters. This function
+    does not mutate the original operator.
+
+    Args:
+        op (.Operator): Operator to update
+        params (Sequence[TensorLike]): New parameters to create operator with
+
+    Returns:
+        .Operator: New operator with updated parameters
+
+    Raises:
+        ValueError: If the shape of the old and new operator parameters don't match
+    """
+    params = _validate_params(op, params)
+
+    return op.__class__(*params, wires=op.wires, **copy.deepcopy(op.hyperparameters))
+
+
+# pylint: disable=missing-docstring
+@bind_new_parameters.register
+def bind_new_parameters_composite_op(
+    op: CompositeOp, params: Optional[Sequence[TensorLike]] = None
+):
+    params = _validate_params(op, params)
+    new_operands = []
+
+    for operand in op.operands:
+        sub_params = params[: operand.num_params]
+        params = params[operand.num_params :]
+        new_operands.append(bind_new_parameters(operand, sub_params))
+
+    return op.__class__(*new_operands)
+
+
+# pylint: disable=missing-docstring
+@bind_new_parameters.register
+def bind_new_parameters_symbolic_op(op: SymbolicOp, params: Optional[Sequence[TensorLike]] = None):
+    params = _validate_params(op, params)
+
+    new_base = bind_new_parameters(op.base, params)
+    new_hyperparameters = copy.deepcopy(op.hyperparameters)
+    _ = new_hyperparameters.pop("base")
+
+    if isinstance(op, Adjoint):
+        # Need this branch because using the other class signature results in a call to
+        # `Adjoint.__new__` which doesn't raise an error but does return an unusable
+        # object
+        return Adjoint(new_base)
+
+    return op.__class__(new_base, **new_hyperparameters)
+
+
+@bind_new_parameters.register
+def bind_new_parameters_scalar_symbolic_op(op: ScalarSymbolicOp, params: Optional[Sequence[TensorLike]] = None):
+    params = _validate_params(op, params)
+
+    if isinstance(op, Pow):
+        new_scalar = op.scalar
+        new_base = bind_new_parameters(op.base, params)
+
+        return Pow(new_base, new_scalar)
+
+    new_scalar = params[0]
+    params = params[1:]
+
+    new_base = bind_new_parameters(op.base, params)
+    new_hyperparameters = copy.deepcopy(op.hyperparameters)
+    _ = new_hyperparameters.pop("base")
+
+    try:
+        # `try-except` block to accomodate for the fact that different `ScalarSymbolicOp`
```

Comment:
I think it's just SProd, which I also think does not need the hyperparameters, so I'd probably just make a whole new dispatch for it